cou/steps/plan.py

```diff
@@ -119,3 +128,66 @@ async def create_upgrade_group(
             raise
 
     return group_upgrade_plan
+
+
+def determine_upgrade_target(
+    current_os_release: Optional[OpenStackRelease], current_series: Optional[str]
+) -> str:
+    """Determine the target release to upgrade to.
+
+    Inform user if the cloud is already at the highest supporting release of the current series.
+    :param current_os_release: The current minimum OS release in cloud.
+    :type current_os_release: Optional[OpenStackRelease]
+    :param current_series: The current minimum Ubuntu series in cloud.
+    :type current_series: Optional[str]
+    :raises NoTargetError: When cannot find target to upgrade.
+    :raises HighestReleaseAchieved: When the highest possible OpenStack release is
+    already achieved.
+    :raises ReleaseOutOfRange: When the release is out of the current supporting range.
+    :return: The target OS release to upgrade the cloud to.
+    :rtype: str
+    """
+    if not current_os_release:
+        raise NoTargetError(
+            "Cannot determine the current OS release in the cloud. "
+            "Is this a valid OpenStack cloud?"
+        )
+
+    supporting_lts_series = ", ".join(LTS_TO_OS_RELEASE)
+    if not current_series:
+        raise NoTargetError(
+            "Cannot determine the current Ubuntu series of the cloud series. "
+            f"The supporting series are: {supporting_lts_series}"
+        )
+
+    try:
+        # Check if the release is the "last" supported by the series
+        if str(current_os_release) == LTS_TO_OS_RELEASE[current_series][-1]:
+            raise HighestReleaseAchieved(
+                f"The cloud is already at the latest OpenStack release '{current_os_release}' "
+                f"compatible with series '{current_series}', and COU does not support series "
+                "upgrade. Please manually upgrade series and run COU again."
+            )
+
+        # get the next release as the target from the current cloud os release
+        target = current_os_release.next_release
+        if not target:
+            raise NoTargetError(
+                "Cannot find target to upgrade. Current minimum OS release is "
+                f"'{str(current_os_release)}'. Current Ubuntu series is '{current_series}'."
+            )
+        # raise exception if the upgrade is not in range from ussuri to yoga
+        if current_os_release < OpenStackRelease("ussuri") or OpenStackRelease(
+            target
+        ) > OpenStackRelease("yoga"):
+            raise ReleaseOutOfRange(
+                "At the moment COU only supports upgrade in the range of `ussuri` and `yoga` "
+                "(inclusive) for `focal` series. Current minimum OS release is "
+                f"'{str(current_os_release)}'. Current Ubuntu series is '{current_series}'."
+            )
+        return target
+    except KeyError:  # raise exception if the current series is not in LTS_TO_OS_RELEASE map
```

Comment:
IMO this try and except `KeyError` is catching too much code. Is  it possible to isolate where you are really checking that?